### PR TITLE
8323746: Add PathElement hashCode and equals

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to implement `hashCode()` and `equals()` methods for implementations of `PathElement`.

In doing so, the previous `PathElementImpl` was removed and replaced in favor of distinct `record` implementations, each reflecting its own path element selection type. This also allowed the `PathKind` to be removed as this piece of information is now carried in the sealed type hierarchy. 

It is worth noting, the implementations resides in the `jdk.internal` package and consequently, they are not exposed to clients. So, we could use pattern matching (for example) internally but not in client code.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323746](https://bugs.openjdk.org/browse/JDK-8323746): Add PathElement hashCode and equals (**Enhancement** - P4)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [5685e31a](https://git.openjdk.org/jdk/pull/17651/files/5685e31a59debd5c2fba09671869eed78638b399)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17651/head:pull/17651` \
`$ git checkout pull/17651`

Update a local copy of the PR: \
`$ git checkout pull/17651` \
`$ git pull https://git.openjdk.org/jdk.git pull/17651/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17651`

View PR using the GUI difftool: \
`$ git pr show -t 17651`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17651.diff">https://git.openjdk.org/jdk/pull/17651.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17651#issuecomment-1919075357)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 1]
```diff
@@ -363,54 +363,144 @@ private String breadcrumbs() {
...
+        }
+
+        @Override
+        public String toString() {
+            return "sequenceElement(" + start + ", " + step + ")";
+        }
+
+    }
+
+    public static final class SequenceElement
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:454]: Why are these not empty records?

Comment (minborg) [reply src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:454]: I had that in the beginning but converted them to regular classes as they do not have any record components. But maybe it is better to make them all records for consistency.

Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:454]: I think so, even less code :-)

**Review by mcimadamore — APPROVED** [review]

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 2]
```diff
@@ -363,54 +363,144 @@ private String breadcrumbs() {
...
+
+        @Override
+        public LayoutPath apply(LayoutPath layoutPath) {
+            return layoutPath.sequenceElement();
+        }
+
+        @Override
+        public int hashCode() {
+            return 31;
+        }
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:468]:
```suggestion
```

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java [hunk 3]
```diff
@@ -363,54 +363,144 @@ private String breadcrumbs() {
...
-            return pathOp.apply(layoutPath);
+            return layoutPath.derefElement();
         }
 
-        public PathKind kind() {
-            return kind;
+        @Override
+        public int hashCode() {
+            return 63;
         }
```
Comment (ExE-Boss) [inline src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java:494]:
```suggestion
```

**Review by ExE-Boss — COMMENTED** [review]
These can now be removed:

File: src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java (hunk 3 above)
Comment (minborg) [reply src/java.base/share/classes/jdk/internal/foreign/LayoutPath.java]: So, a record with no component will have a `hashCode()` of zero.  I didn't want the two singleton variants to have the same hash code. I've added a test for this and will remove one of the overrides.

**Pull request closed without merging**